src/pip/_internal/index.py

```diff
@@ -226,6 +226,119 @@ def _get_html_page(link, session=None):
+    pass
+
+
+class _BinaryLinkNotPermitted(_LinkParseError):
+    pass
+
+
+class _MacOSXZip(_LinkParseError):
+    pass
+
+
+class _IncorrectWheelName(_LinkParseError):
+    pass
+
+
+class _IncompatibleWheel(_LinkParseError):
+    pass
+
+
+class _SourceLinkNotPermitted(_LinkParseError):
+    pass
+
+
+class _PythonVersionNotSupported(_LinkParseError):
+    pass
+
+
+def _egg_info_matches(egg_info, canonical_name):
+    """Pull the version part out of a string.
+
+    :param egg_info: The string to parse. E.g. foo-2.1
+    :param canonical_name: The canonicalized name of the package this
+        belongs to.
+    """
+    try:
+        version_start = _find_name_version_sep(egg_info, canonical_name) + 1
+    except ValueError:
+        return None
+    version = egg_info[version_start:]
+    if not version:
+        return None
+    return version
+
+
+_PYTHON_VERSION_RE = re.compile(r'-py([123]\.?[0-9]?)$')
+
+
+def _parse_package_version_non_wheel(egg_info, search):
+    # This should be up by the search.ok_binary check, but see issue 2700.
+    if "source" not in search.formats:
+        raise _SourceLinkNotPermitted
+
+    info = _egg_info_matches(egg_info, search.canonical)
+    if info is None:
+        raise _LinkParseError
+    match = _PYTHON_VERSION_RE.search(info)
+    if not match:
+        return info
+    if match.group(1) != sys.version[:3]:
+        raise _PythonVersionNotSupported
+    return info[:match.start()]
+
+
+def _parse_package_version(link, search, valid_tags):
+    """Parse package version from link, while filtering invalid inputs.
+
+    Returns a string representing the package version, or None if a version
+    cannot be parsed.
+
+    May raise a variety of `_LinkParseError` subclasses to indicate failures
+    during the process:
+
+    * `_LinkNotAFile`: The link does not look like a file.
+    * `_LinkFormatUnsupported`: The link has an unsupported extension.
+    * `_BinaryLinkNotPermitted`: The link has a supported extension, but it is
+        a binary, and the user explicitly refuses them.
+    * `_MacOSXZip`: The link is a ZIP built for Mac OS X. I don't know why this
+        is an error. Historical reasons?
+    * `_IncorrectWheelName`: The wheel's project name is not what we are
+        looking for (indicated by the `search` argument).
+    * `_IncompatibleWheel` The wheel is not compatible with the current
+        platform (indicated by the `valid_tags` argument).
+    """
+    if link.egg_fragment:
+        return _parse_package_version_non_wheel(link.egg_fragment, search)
+    egg_info, ext = link.splitext()
+    if not ext:
+        raise _LinkNotAFile
+    if ext not in SUPPORTED_EXTENSIONS:
+        raise _LinkFormatUnsupported
+    if "binary" not in search.formats and ext == wheel_ext:
+        raise _BinaryLinkNotPermitted
+    if "macosx10" in link.path and ext == '.zip':
+        raise _MacOSXZip
+    if ext != wheel_ext:
+        return _parse_package_version_non_wheel(egg_info, search)
```

Comment:
It seems like it would make for a nicer flow if there was just one call to `_parse_package_version_non_wheel()` to go through (similar to the original code). What would a PR preserving that aspect look like?